client/rpc/src/author/mod.rs

```diff
@@ -52,33 +52,147 @@ pub struct Author<P, Client> {
+		})?;
+
+		ctx_module.register_method::<_, Bytes>("author_rotateKeys", |params, author| {
+			log::info!("author_rotateKeys [{:?}]", params);
+			author.deny_unsafe.check_if_safe()?;
+
+			let best_block_hash = author.client.info().best_hash;
+			author.client.runtime_api().generate_session_keys(
+				&generic::BlockId::Hash(best_block_hash),
+				None,
+			)
+			.map(Into::into)
+			.map_err(|api_err| Error::Client(Box::new(api_err)).into())
+		})?;
+
+		ctx_module.register_method("author_hasSessionKeys", |params, author| {
+			log::info!("author_hasSessionKeys [{:?}]", params);
+			author.deny_unsafe.check_if_safe()?;
+
+			let session_keys: Bytes = params.one()?;
+			let best_block_hash = author.client.info().best_hash;
+			let keys = author.client.runtime_api().decode_session_keys(
+				&generic::BlockId::Hash(best_block_hash),
+				session_keys.to_vec(),
+			).map_err(|e| RpseeCallError::Failed(Box::new(e)))?
+			.ok_or_else(|| Error::InvalidSessionKeys)?;
+
+			Ok(SyncCryptoStore::has_keys(&*author.keystore, &keys))
+		})?;
+
+		ctx_module.register_method("author_hasKey", |params, author| {
+			log::info!("author_hasKey [{:?}]", params);
+			author.deny_unsafe.check_if_safe()?;
+
+			// TODO: this compiles, but I don't know how it could actually work...?
+			// let (public_key, key_type)  = params.parse::<(Vec<u8>, KeyTypeId)>()?;
+			let (public_key, key_type)  = params.parse::<(Vec<u8>, String)>()?;
+			let key_type = key_type.as_str().try_into().map_err(|_| Error::BadKeyType)?;
+			Ok(SyncCryptoStore::has_keys(&*author.keystore, &[(public_key, key_type)]))
+		})?;
+
+		ctx_module.register_method::<_, TxHash<P>>("author_submitExtrinsic", |params, author| {
+			log::info!("author_submitExtrinsic [{:?}]", params);
+			// TODO: make is possible to register async methods on jsonrpsee servers.
+			//https://github.com/paritytech/jsonrpsee/issues/291
+			//
+			// NOTE(niklasad1): will block the connection task on the server.
+			let ext: Bytes = params.one()?;
+			let xt = match Decode::decode(&mut &ext[..]) {
+				Ok(xt) => xt,
+				Err(err) => return Err(RpseeCallError::Failed(err.into())),
+			};
+			let best_block_hash = author.client.info().best_hash;
+			let fut = author.pool.submit_one(&generic::BlockId::hash(best_block_hash), TX_SOURCE, xt);
+
+			futures::executor::block_on(fut)
+				.map_err(|e| e.into_pool_error()
+					.map(|e| RpseeCallError::Failed(Box::new(e)))
+					.unwrap_or_else(|e| RpseeCallError::Failed(Box::new(e))))
+		})?;
+
+		ctx_module.register_method::<_, Vec<Bytes>>("author_pendingExtrinsics", |_, author| {
+			log::info!("author_pendingExtrinsics");
+			Ok(author.pool.ready().map(|tx| tx.data().encode().into()).collect())
+		})?;
+
+		ctx_module.register_method::<_, Vec<TxHash<P>>>("author_removeExtrinsic", |params, author| {
+			log::info!("author_removeExtrinsic [{:?}]", params);
+			author.deny_unsafe.check_if_safe()?;
+
+			let bytes_or_hash: Vec<hash::ExtrinsicOrHash<TxHash<P>>> = params.parse()?;
```

Comment:
@niklasad1 I ran into a problem when parsing the params here, the compiler complaining about the type not being `Deserialize`. 
The `hash::ExtrinsicOrHash<TxHash<P>>` comes from [here](https://github.com/paritytech/substrate/blob/20f1e369edb860a8287d79dd85a27b50e391b07c/client/rpc-api/src/author/hash.rs#L29) which derives `Deserialize`, but the type parameter, `TxHash<P>` in our case, is not as you can see [here](https://github.com/paritytech/substrate/blob/master/primitives/transaction-pool/src/pool.rs#L180) so to get around that I added `DeserializeOwned` [here](https://github.com/paritytech/substrate/blob/20f1e369edb860a8287d79dd85a27b50e391b07c/primitives/transaction-pool/src/pool.rs#L180).

I guess the question I have is: how could it work before? Like, how can `jsonrpc` deserialize the params when `jsonrpsee` needed the extra bound?

Reply:
Hmm, could be. Do you know if there are such custom bounds in these types? I couldn't find any. :/

Reply:
Yeah, I realized what's difference is.

In `jsonrpc` the Params type is `derive(Deserialize)` where the type bounds are handwritten in serde.
In `jsonrpsee` the params explicitly has the `parse()` which requires the type to be `Deserialize/DeserializeOwned` that is called in closure.

@dvdplm, k?

Reply:
Lovely, that makes sense. I guess we'll see if adding the bound like I had to do becomes a problem or not.